### PR TITLE
Merge in upstream main for 5.2 merge

### PR DESCRIPTION
**Pull request by ncik-roberts**
Merges the HEAD of the parent PR #81 with a reasonably recent `upstream/main`. Fix all tests.

**Caveat:** I probably should have done the merge with `upstream/v5.1-502`, which is the tagged version for 5.2. I skimmed through the commits to `main` since then (there are about 20) and almost all of them are bugfixes. There is a new feature: https://github.com/ocaml/merlin/pull/1745, but I'm not worried about it as it doesn't touch existing code paths.

Note for next version bump: be sure to merge with the tagged version! I don't think we hit any difficulty in this merge, but that's half by luck.

This merge was difficult. The last merge was done with https://github.com/ocaml/merlin/tree/v4.14-501, which does not share history with upstream `main` (or `v5.1-502`). This means that `git merge` creates a lot of spurious conflicts.

The way I created this PR was as follows:
  1. I created a parallel branch `B` where I checked out to `v4.14-501`, and committed a change bringing the files to the same state as `upstream/main`.
  2. I then merged `B` with the HEAD of #81.
  4. I returned to the branch of the present PR, and merged with `upstream/main` to make the branches share history (https://github.com/janestreet/merlin-jst/pull/82/commits/6b3f6af65d1088e1c7d66aec170b90ed1a50427e).
  5. I then made two further commits: (i) a commit making files look like `B` after step 2 (https://github.com/janestreet/merlin-jst/pull/82/commits/240f8aac71b9394828871cbaee55cb0f02639a8f), and (ii) a series of commits fixing up the conflicts and fixing the tests.

I expect that future merges will be less painful &mdash; I think we'll just be able to run `git merge` with the tagged version.

The reason I went to all this effort is because it halved the number of conflicts. There are about 400 conflicts in `B`, and about 800 conflicts that I would have had to fix if I had just merged with `upstream/main` from the start.

There might have been a better way to do this. Let me know if you think of one.

### Instructions for reviewer

Review this diff: https://github.com/janestreet/merlin-jst/pull/82/files/6b3f6af65d1088e1c7d66aec170b90ed1a50427e..150f7876a9dbf4c50674be292a330b2b5df0afd3

This is the diff between:
  * The conflicts created by merging with `upstream/main`
  * The final resolution.

Explanation of three-way diff:
  * The top third is **the version of 5.2 merlin as of #81**.
  * The middle third is...I don't really know. I really didn't look at this. It's just the auto-detected merge base, which isn't useful here.
  * The bottom third is **upstream merlin as of a recent-ish upstream main** (just after v5.1-502).

I usually take the upper third of the diff. Places where I don't are notable.

TODO: We should scan through the diff between merlin-jst's implementation of type-checking and flambda-backend's.

### Notable diffs
* `src/ocaml/typing/typecore.ml` drops upstream's approach for recovery when typing functions. Our approach is slightly different. We should make them converge, but this PR is not the place to do it: it's tricky, and there's a lot going on in this PR already.
* The lack of diff in the overall PR for occurrences-related files is important. @liam923 says this is the right outcome.

### CODE REVIEW

File: src/ocaml/merlin_specific/browse_raw.ml [hunk 1]
```diff
@@ -346,48 +340,14 @@ let of_method_call obj meth loc =
...
-     allows us to respond with the *parameter*'s type (as opposed to the
-     function's type) when the user queries the label:
-
-     let f ?y:(x = 3) () = x
-           ^
-  *)
-  match param.fp_kind with
-  | Tparam_pat pat -> of_pattern pat
-  | Tparam_optional_default (pat, expr, _) ->
-      of_pattern pat ** of_expression expr
```
Comment (liam923) [inline src/ocaml/merlin_specific/browse_raw.ml:354]: I don't know the story behind this change - it looks like you made this a few months before I started - but it looks like you didn't preserve it. Was that intentional?

If the answer is yes, I trust that it's fine. I just want to make sure it wasn't a mistake.

File: src/ocaml/merlin_specific/tast_helper.ml [hunk 2]
```diff
@@ -14,15 +14,8 @@ module Pat = struct
       | None -> str.Asttypes.loc
       | Some loc -> loc
     in
-<<<<<<< HEAD
     let mode = Mode.Value.newvar () in
-    let uid = Uid.mk ~current_unit:(Env.get_unit_name ()) in
```
Comment (liam923) [inline src/ocaml/merlin_specific/tast_helper.ml:18]: Do you know the story behind this line? It looks like it was added in the 4.14.1-19 merge, but was never added upstream.

File: src/kernel/mconfig.ml [hunk 3]
```diff
@@ -139,15 +133,10 @@ let dump_merlin x =
...
-<<<<<<< HEAD
-    "unit_name"    , Json.option Json.string x.unit_name;
-    "wrapping_prefix" , Json.option Json.string x.wrapping_prefix;
-||||||| 7b73c6aa3f
-=======
     "source_root"  , Json.option Json.string x.source_root;
     "unit_name"    , Json.option Json.string x.unit_name;
     "wrapping_prefix" , Json.option Json.string x.wrapping_prefix;
->>>>>>> upstream/main
+    "source_root"  , Json.option Json.string x.source_root;
```
Comment (liam923) [inline src/kernel/mconfig.ml]: `source_root` dup

File: src/kernel/mconfig.ml [hunk 4]
```diff
@@ -695,11 +618,6 @@ let ocaml_ignored_parametrized_flags = [
   "-zero-alloc-checker-details-cutoff";
   "-zero-alloc-checker-join";
   "-dgranularity";
-||||||| 7b73c6aa3f
-  "-error-style"; "-dump-dir";
-=======
-  "-error-style"; "-dump-dir"; "-cmi-file";
```
Comment (liam923) [inline src/kernel/mconfig.ml]: I think we're missing `-cmi-file`?

**Review by liam923 — CHANGES_REQUESTED** [review]
Mostly looks good to me

File: tests/test-dirs/type-enclosing/underscore-ids.t [hunk 5]
```diff
@@ -344,40 +343,13 @@ We try several places in the identifier to check the result stability
...
-          "col": 17
-        },
-        "type": "'a",
-        "tail": "no"
-      },
-      {
-        "start": {
-          "line": 2,
-=======
->>>>>>> upstream/main
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply tests/test-dirs/type-enclosing/underscore-ids.t]: Despite what this diff would imply, this test output is actually unchanged from this PR chain.
https://github.com/janestreet/merlin-jst/commit/6b3f6af65d1088e1c7d66aec170b90ed1a50427e#diff-926789fce943b27f668ecb45886e79492a38e6240562ef67587f908cbc22acf1

I think it has something to do with the way that we handle error recovery for n-ary functions internally. We should try to align these at some point. But I think things are OK.

File: src/frontend/query_commands.ml [hunk 6]
```diff
@@ -851,22 +838,12 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
...
           find_in_path_normalized (Mconfig.source_path config @ Mconfig.hidden_source_path config) x
-||||||| 7b73c6aa3f
-          find_in_path_uncap (Mconfig.source_path config) x
-=======
-          find_in_path_normalized (Mconfig.source_path config) x
->>>>>>> upstream/main
         with Not_found -> try
-<<<<<<< HEAD
+            (* CR nroberts: upstream doesn't use hidden here *)
             find_in_path_normalized (Mconfig.build_path config @ Mconfig.hidden_build_path config) x
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply src/frontend/query_commands.ml]: It seems strange that `source_path` and `build_path` differs in this way.

File: src/frontend/query_commands.ml [hunk 7]
```diff
@@ -851,22 +838,12 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
     let rec aux = function
       | [] -> raise Not_found
       | x :: xs ->
+          (* CR nroberts: upstream doesn't use hidden here *)
         try
-<<<<<<< HEAD
           find_in_path_normalized (Mconfig.source_path config @ Mconfig.hidden_source_path config) x
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply src/frontend/query_commands.ml]: OK, changed.

File: src/ocaml/merlin_specific/browse_raw.ml (hunk 1 above)
Comment (ncik-roberts) [reply src/ocaml/merlin_specific/browse_raw.ml:354]: The code appears elsewhere in the file. Keeping it was intentional:

```ocaml
(* We should consider taking into account param.fp_loc at some point, as it
   allows us to respond with the *parameter*'s type (as opposed to the
   function's type) when the user queries the label:

   let f ?y:(x = 3) () = x
         ^
*)
and of_function_param fp = of_function_param_kind fp.fp_kind

and of_function_param_kind = function
  | Tparam_pat pat -> of_pattern pat
  | Tparam_optional_default (pat, exp, _) ->
    of_pattern pat ** of_expression exp
```

File: src/ocaml/merlin_specific/tast_helper.ml (hunk 2 above)
Comment (ncik-roberts) [reply src/ocaml/merlin_specific/tast_helper.ml:18]: Upstream added `uid` as a parameter to `var`. That's why I can delete this line now.

File: src/kernel/mconfig.ml (hunk 4 above)
Comment (ncik-roberts) [reply src/kernel/mconfig.ml]: I don't think so. Merlin handles that flag now:

```
  ( "-cmi-file",
    Marg.param "file" (fun cmi_file ocaml ->
        {ocaml with cmi_file = Some cmi_file}),
    "<file>  Use the <file> interface to type-check"
  );

```

**Comment by ncik-roberts** [comment]
Thanks for the review. It should be back to you @liam923

File: src/frontend/query_commands.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (liam923) [reply src/frontend/query_commands.ml]: Agreed, I'm not sure what the reasoning is. I think we can merge this as is, and I can ask upstream devs about it.

File: src/ocaml/merlin_specific/browse_raw.ml (hunk 1 above)
Comment (liam923) [reply src/ocaml/merlin_specific/browse_raw.ml:354]: Ah I missed that. Looks good then

File: src/kernel/mconfig.ml (hunk 4 above)
Comment (liam923) [reply src/kernel/mconfig.ml]: Ah ok

File: tests/test-dirs/type-enclosing/underscore-ids.t (hunk 5 above)
[the remark this answers is not in the document]
Comment (liam923) [reply tests/test-dirs/type-enclosing/underscore-ids.t]: Ok yes I agree this is fine

File: .github/dependabot.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+version: 2
+updates:
+  - package-ecosystem: github-actions
+    directory: /
+    schedule:
+      interval: weekly
+    labels:
+      - no changelog
```
Comment (liam923) [inline .github/dependabot.yml:8]: I think the changes to the `.github` folder should be reverted?

**Review by liam923 — APPROVED** [review]
Other than the CI stuff, things look good to me

File: .github/dependabot.yml (hunk 8 above)
Comment (ncik-roberts) [reply .github/dependabot.yml]: Good idea, done.

**Comment by ncik-roberts** [comment]
I'll leave instructions for how to merge this.

**Pull request merged**